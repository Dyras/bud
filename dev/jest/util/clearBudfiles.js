const {noop} = require('lodash')
const rimraf = require('rimraf')

module.exports = async () => {
  return rimraf(
    `${process.cwd()}/examples/*/.budfiles`,
    {
      glob: true,
    },
    noop,
  )
}
