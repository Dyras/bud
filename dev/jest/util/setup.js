/* eslint-disable no-console */
// @ts-check

const chalk = require('chalk')
const execa = require('execa')

const clearArtifacts = require('./clearArtifacts')
const paths = require('./paths')

const task = async (cmd, paths) => {
  console.log(chalk.blue`\nyarn bud ${cmd}\n`)

  await Promise.all(
    paths.map(async ex => {
      const task = execa.command(`yarn bud ${cmd}`, {
        cwd: ex.cwd,
      })
      task.stdout.pipe(process.stdout)
      await task
    }),
  )
}

module.exports = async () => {
  await clearArtifacts()

  const examples = await paths()

  await task('build --no-dashboard', examples)
}
