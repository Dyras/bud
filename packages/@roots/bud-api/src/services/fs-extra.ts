import {fs} from '@roots/bud-support'

export const ensureDirSync = fs.ensureDirSync
export const pathExistsSync = fs.pathExistsSync
