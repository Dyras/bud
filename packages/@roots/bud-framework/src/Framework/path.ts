import {join} from 'path'

import {Framework} from '..'

interface path {
  (
    this: Framework,
    key: keyof Framework.Locations,
    ...path: string[]
  ): string
}

function path(
  key: keyof Framework.Locations,
  ...path: string[]
): string {
  return join(
    ...[
      key !== 'project'
        ? this.hooks.filter('location/project')
        : false,
      this.hooks.filter(`location/${key}`),
      ...(path ?? []),
    ].filter(Boolean),
  )
}

export {path}
