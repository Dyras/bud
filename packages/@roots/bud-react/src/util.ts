/**
 * @module @roots/bud-react
 */

import type {Configuration} from 'webpack'

/**
 * @function addRefresh
 */
function addRefresh(entries, [name, assets]) {
  return {
    ...(entries ?? {}),
    [name]: {
      ...assets,
      import: [
        'react-refresh/runtime',
        ...(assets.import ?? []),
      ],
    },
  }
}

/**
 * @function reducer
 */
function reducer(entry: Configuration['entry']) {
  return entry
    ? Object.entries(entry).reduce(addRefresh, entry)
    : {}
}

/**
 * @exports default
 */
export default reducer
