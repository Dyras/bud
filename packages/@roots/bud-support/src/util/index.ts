export {notify} from './notify'
export {dump} from './dump'
export {format} from './format'
export {processHandler} from './processHandler'

import InterpolateHtmlPlugin from './InterpolateHtmlPlugin'
export {InterpolateHtmlPlugin}
