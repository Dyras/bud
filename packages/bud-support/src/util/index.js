"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterpolateHtmlPlugin = exports.formatWebpackMessages = exports.eslintFormatter = exports.processHandler = exports.format = exports.dump = exports.notify = void 0;
var notify_1 = require("./notify");
Object.defineProperty(exports, "notify", { enumerable: true, get: function () { return notify_1.notify; } });
var dump_1 = require("./dump");
Object.defineProperty(exports, "dump", { enumerable: true, get: function () { return dump_1.dump; } });
var format_1 = require("./format");
Object.defineProperty(exports, "format", { enumerable: true, get: function () { return format_1.format; } });
var processHandler_1 = require("./processHandler");
Object.defineProperty(exports, "processHandler", { enumerable: true, get: function () { return processHandler_1.processHandler; } });
const eslintFormatter_1 = __importDefault(require("./eslintFormatter"));
exports.eslintFormatter = eslintFormatter_1.default;
const InterpolateHtmlPlugin_1 = __importDefault(require("./InterpolateHtmlPlugin"));
exports.InterpolateHtmlPlugin = InterpolateHtmlPlugin_1.default;
const formatWebpackMessages_1 = __importDefault(require("./formatWebpackMessages"));
exports.formatWebpackMessages = formatWebpackMessages_1.default;
//# sourceMappingURL=index.js.map