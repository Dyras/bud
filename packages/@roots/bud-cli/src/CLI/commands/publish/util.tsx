import {chalk, globby} from '@roots/bud-support'
import {dirname} from 'path'

/**
 * Publishable matches (glob pattern)
 */
export const publishables = [
  ...globby
    .sync([
      '**/node_modules/bud-*/publish/*',
      '**/node_modules/*/bud-*/publish/*',
    ])
    .filter(
      (value, index, self) => self.indexOf(value) === index,
    ),
]

export const formatted = publishables.reduce((a, file) => {
  const ext = dirname(dirname(file)).replace('node_modules/', '')

  const template = file
    .replace(`${ext}/publish/`, '')
    .replace('node_modules/', '')

  return `${a} \n${chalk.blue(ext)} ${chalk.green(template)}`
}, ``)
