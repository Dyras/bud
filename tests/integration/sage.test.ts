import {readJson} from 'fs-extra'
import {join} from 'path'

import {Assets, Entrypoints, helper} from '../util/integration'

const suite = helper(
  'sage',
  'examples/sage',
  'public',
  '/app/themes/sage/public/',
)

jest.setTimeout(1000000)

describe(suite.name, () => {
  let assets: Assets
  let entrypoints: Entrypoints

  beforeAll(async () => {
    assets = await suite.setup()
    entrypoints = await readJson(
      suite.distPath('entrypoints.json'),
    )

    return
  })

  describe('entrypoints.json', () => {
    it('has expected app entries', () => {
      expect(entrypoints.app.js).toBeInstanceOf(Array)
      expect(entrypoints.app.js).toHaveLength(2)
      expect(entrypoints.app.css).toBeInstanceOf(Array)
      expect(entrypoints.app.css).toHaveLength(1)
      expect(entrypoints.app.dependencies).toEqual([])
    })

    it('has expected editor entries', () => {
      expect(entrypoints.editor.js).toBeInstanceOf(Array)
      expect(entrypoints.editor.js).toHaveLength(2)
      expect(entrypoints.editor.css).toBeInstanceOf(Array)
      expect(entrypoints.editor.css).toHaveLength(1)
      expect(entrypoints.editor.dependencies).toEqual([
        'wp-edit-post',
        'wp-dom-ready',
        'wp-blocks',
      ])
    })

    it('has expected customizer entries', () => {
      expect(entrypoints.customizer.js).toBeInstanceOf(Array)
      expect(entrypoints.customizer.js).toHaveLength(2)
      expect(entrypoints.customizer.dependencies).toEqual([
        'jquery',
      ])
    })
  })

  describe('runtime', () => {
    it('has contents', () => {
      expect(assets['runtime.js'].length).toBeGreaterThan(10)
    })
    it('is transpiled', () => {
      expect(assets['runtime.js'].includes('import')).toBeFalsy()
    })
  })

  describe('app', () => {
    it('has contents', () => {
      expect(assets['app.js'].length).toBeGreaterThan(10)
    })
    it('is transpiled', () => {
      expect(assets['app.js'].includes('import')).toBeFalsy()
    })
    it('css: has contents', () => {
      expect(assets['app.css'].length).toBeGreaterThan(10)
    })
    it('css: is transpiled', () => {
      expect(assets['app.css'].includes('@import')).toBe(false)
    })
    it('css: @tailwind directive is transpiled', () => {
      expect(assets['app.css'].includes('@apply')).toBe(false)
    })
    it('css: has whitespace removed', () => {
      expect(assets['app.css'].match(/    /)).toBeFalsy()
    })
    it('css: has breaks removed', () => {
      expect(assets['app.css'].match(/\\n/)).toBeFalsy()
    })
  })

  describe('editor', () => {
    it('has contents', () => {
      expect(assets['editor.js'].length).toBeGreaterThan(10)
    })
    it('is transpiled', () => {
      expect(assets['editor.js'].includes('import')).toBeFalsy()
    })
    it('css: has contents', () => {
      expect(assets['editor.css'].length).toBeGreaterThan(10)
    })
    it('css: is transpiled', () => {
      expect(assets['editor.css'].includes('@import')).toBe(
        false,
      )
    })
    it('css: @tailwind directive is transpiled', () => {
      expect(assets['editor.css'].includes('@apply')).toBe(false)
    })
    it('css: has whitespace removed', () => {
      expect(assets['editor.css'].match(/    /)).toBeFalsy()
    })
    it('css: has breaks removed', () => {
      expect(assets['editor.css'].match(/\\n/)).toBeFalsy()
    })
  })

  describe('customizer', () => {
    it('has contents', () => {
      expect(assets['customizer.js'].length).toBeGreaterThan(10)
    })
    it('is transpiled', () => {
      expect(
        assets['customizer.js'].includes('import'),
      ).toBeFalsy()
    })
  })

  describe('snapshots', () => {
    it('package.json is unchanged', async () => {
      const artifact = await readJson(
        join(process.cwd(), 'examples/sage/package.json'),
      )

      expect(artifact).toMatchSnapshot()
    })

    it('public/manifest.json matches expectations', async () => {
      const artifact = await readJson(
        join(
          process.cwd(),
          'examples/sage/public/manifest.json',
        ),
      )

      expect(Object.entries(artifact).length).toEqual(8)

      expect(
        Object.values<string>(artifact).filter(
          v =>
            (v.startsWith('assets/') ||
              v.startsWith('images/') ||
              v.startsWith('app') ||
              v.startsWith('editor') ||
              v.startsWith('runtime') ||
              v.startsWith('customizer')) &&
            (v.endsWith('.js') ||
              v.endsWith('.css') ||
              v.endsWith('.jpeg')),
        ).length,
      ).toBe(Object.values(artifact).length)

      expect(artifact['app.js']).toMatch(/app\.[\d|\w]*\.js/)
      expect(artifact['app.css']).toMatch(/app\.[\d|\w]*\.css/)
      expect(artifact['editor.js']).toMatch(
        /editor\.[\d|\w]*\.js/,
      )
      expect(artifact['editor.css']).toMatch(
        /editor\.[\d|\w]*\.css/,
      )
      expect(artifact['customizer.js']).toMatch(
        /customizer\.[\d|\w]*\.js/,
      )
      expect(artifact['runtime.js']).toMatch(
        /runtime\.[\d|\w]*\.js/,
      )

      expect(Object.keys(artifact)).toMatchSnapshot()
    })

    it('.budfiles/bud.webpack.config.js', async () => {
      const artifact = await import(
        join(
          process.cwd(),
          'examples/sage/storage/bud/bud.webpack.config.js',
        )
      ).then(artifact => artifact())

      expect(artifact.name).toMatchSnapshot()
      expect(artifact.entry).toMatchSnapshot()
      expect(artifact.mode).toMatchSnapshot()
      expect(artifact.optimization).toMatchSnapshot()
      expect(artifact.bail).toMatchSnapshot()
      expect(artifact.cache).toMatchSnapshot()
    })

    it('module map matches snapshot', async () => {
      const artifact = await readJson(
        join(
          process.cwd(),
          'examples/sage/storage/bud/bud-modules.json',
        ),
      )

      expect(artifact.chunks).toMatchSnapshot({
        byName: {
          app: expect.any(Number),
          customizer: expect.any(Number),
          editor: expect.any(Number),
          runtime: expect.any(Number),
        },
        bySource: {
          '0 app': expect.any(Number),
          '0 customizer': expect.any(Number),
          '0 editor': expect.any(Number),
          '1 app': expect.any(Number),
          '1 customizer': expect.any(Number),
          '1 editor': expect.any(Number),
        },
        usedIds: [
          expect.any(Number),
          expect.any(Number),
          expect.any(Number),
          expect.any(Number),
        ],
      })
    })
  })
})
