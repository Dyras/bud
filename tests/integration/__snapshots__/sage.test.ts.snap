// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`examples/sage [snapshots] .budfiles/bud.webpack.config.js 1`] = `"bud"`;

exports[`examples/sage [snapshots] .budfiles/bud.webpack.config.js 2`] = `
Object {
  "app": Object {
    "import": Array [
      "scripts/app.js",
      "styles/app.css",
    ],
  },
  "customizer": Object {
    "import": Array [
      "scripts/customizer.js",
    ],
  },
  "editor": Object {
    "import": Array [
      "scripts/editor.js",
      "styles/editor.css",
    ],
  },
}
`;

exports[`examples/sage [snapshots] .budfiles/bud.webpack.config.js 3`] = `"production"`;

exports[`examples/sage [snapshots] .budfiles/bud.webpack.config.js 4`] = `
Object {
  "minimize": true,
  "minimizer": Array [
    "...",
    Object {
      "options": Object {
        "minify": "<<Function 'cssnanoMinify'>>",
        "minimizerOptions": Object {
          "preset": Array [
            "default",
            Object {
              "discardComments": Object {
                "removeAll": true,
              },
            },
          ],
        },
        "parallel": true,
        "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        "warningsFilter": "<<Function 'warningsFilter'>>",
      },
    },
    Object {
      "options": Object {
        "minify": "<<Function 'cssnanoMinify'>>",
        "minimizerOptions": "<<Circular reference to 'config.optimization.minimizer.[1].options.minimizerOptions'>>",
        "parallel": true,
        "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        "warningsFilter": "<<Function 'warningsFilter'>>",
      },
    },
  ],
  "moduleIds": "deterministic",
  "runtimeChunk": Object {
    "name": "<<Function 'name'>>",
  },
  "splitChunks": Object {
    "cacheGroups": Object {
      "bud": Object {
        "chunks": "all",
        "name": "<<Function 'name'>>",
        "priority": -10,
        "reuseExistingChunk": true,
        "test": /\\(\\[\\\\\\\\/\\]@roots\\|webpack\\|style-loader\\|tslib\\|ansi\\|html-entities\\|css-loader\\)/,
      },
      "vendor": Object {
        "chunks": "all",
        "name": "<<Function 'name'>>",
        "priority": -20,
        "reuseExistingChunk": true,
        "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
      },
    },
    "defaultSizeTypes": Array [
      "...",
    ],
  },
}
`;

exports[`examples/sage [snapshots] .budfiles/bud.webpack.config.js 5`] = `true`;

exports[`examples/sage [snapshots] .budfiles/bud.webpack.config.js 6`] = `
Object {
  "buildDependencies": Object {
    "bud": Array [
      "/Volumes/Samsung/Code/roots/bud/examples/sage/bud.config.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/sage/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-preset-wordpress/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-preset-recommend/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-babel/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-entrypoints/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-postcss/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-react/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-wordpress-externals/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-wordpress-dependencies/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-wordpress-manifests/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-eslint/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-prettier/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-stylelint/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/packages/@roots/bud-tailwindcss/lib/cjs/index.js",
      "/Volumes/Samsung/Code/roots/bud/examples/sage/storage/bud/cache/eslint.json",
    ],
  },
  "cacheDirectory": "/Volumes/Samsung/Code/roots/bud/examples/sage/storage/bud/cache",
  "type": "filesystem",
  "version": "zvhz6mgijnd8f0plb5gqyjbj53s_",
}
`;

exports[`examples/sage [snapshots] module map matches snapshot 1`] = `
Object {
  "byName": Object {
    "app": Any<Number>,
    "customizer": Any<Number>,
    "editor": Any<Number>,
    "runtime": Any<Number>,
  },
  "bySource": Object {
    "0 app": Any<Number>,
    "0 customizer": Any<Number>,
    "0 editor": Any<Number>,
    "1 app": Any<Number>,
    "1 customizer": Any<Number>,
    "1 editor": Any<Number>,
  },
  "usedIds": Array [
    Any<Number>,
    Any<Number>,
    Any<Number>,
    Any<Number>,
  ],
}
`;

exports[`examples/sage [snapshots] package.json is unchanged 1`] = `
Object {
  "browserslist": Array [
    "extends @wordpress/browserslist-config",
  ],
  "devDependencies": Object {
    "@roots/bud": "workspace:*",
    "@roots/sage": "workspace:*",
    "@types/eslint": Any<String>,
    "@types/react": Any<String>,
    "@types/react-dom": Any<String>,
    "@wordpress/browserslist-config": Any<String>,
    "eslint": Any<String>,
    "postcss": Any<String>,
    "postcss-import": Any<String>,
    "postcss-preset-env": Any<String>,
    "react": Any<String>,
    "react-dom": Any<String>,
    "tailwindcss": Any<String>,
  },
  "name": "example-sage",
  "private": true,
}
`;

exports[`examples/sage [snapshots] public/manifest.json matches expectations 1`] = `
Array [
  "app.css",
  "app.js",
  "editor.css",
  "editor.js",
  "customizer.js",
  "runtime.js",
  "assets/image.jpeg",
  "images/image.jpeg",
]
`;

exports[`examples/sage package.json is unchanged 1`] = `
Object {
  "browserslist": Array [
    "extends @wordpress/browserslist-config",
  ],
  "devDependencies": Object {
    "@roots/bud": "workspace:*",
    "@roots/sage": "workspace:*",
    "@types/eslint": Any<String>,
    "@types/react": Any<String>,
    "@types/react-dom": Any<String>,
    "@wordpress/browserslist-config": Any<String>,
    "eslint": Any<String>,
    "postcss": Any<String>,
    "postcss-import": Any<String>,
    "postcss-preset-env": Any<String>,
    "react": Any<String>,
    "react-dom": Any<String>,
    "tailwindcss": Any<String>,
  },
  "name": "example-sage",
  "private": true,
}
`;
